File: 搜索/例题/3.dfs暴力搜索/python/main.py
import sys

sys.setrecursionlimit(100000)  # 例如这里设置为十万

flag = False


def dfs(list1, is_visited, n, count, time):
    global flag
    if flag is True:
        return
    if count == n:
        flag = True
        return
    for i in range(n):
        if is_visited[i] is False and time <= list1[i][0] + list1[i][1]:
            is_visited[i] = True
            temp = max(time, list1[i][0])
            dfs(list1, is_visited, n, count + 1, temp + list1[i][2])
            is_visited[i] = False


if __name__ == '__main__':
    t = int(input())
    while t > 0:
        t -= 1
        flag = False
        list1 = []
        n = int(input())
        for i in range(n):
            temp = []
            s = input()
            s = s.split(" ")
            for j in range(3):
                temp.append(int(s[j]))
            list1.append(temp)
        is_visited = [False] * n
        dfs(list1, is_visited, n, 0, 0)
        if flag is True:
            print("YES")
        else:
            print("NO")

File: 搜索/例题/3.dfs暴力搜索/java/Main.java
import java.util.*;
public class Main {

    static boolean flag;

    static void dfs(List<int[]> list,boolean[] isVisited,int n,int count,int time){
        if(flag){
            return;
        }
        if(count==n){
            flag = true;
            return;
        }
        for(int i=0;i<n;i++){
            if(!isVisited[i]&&time<=list.get(i)[0]+list.get(i)[1]){
                isVisited[i]=true;
                int temp = Math.max(time,list.get(i)[0]);
                dfs(list,isVisited,n,count+1,temp+list.get(i)[2]);
                isVisited[i]=false;
            }
        }
    }
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int t = scanner.nextInt();
        while (t>0){
            t--;
            flag = false;
            int n = scanner.nextInt();
            List<int[]> list = new ArrayList<>();
            for(int i=0;i<n;i++){
                int[] temp = new int[3];
                for(int j=0;j<3;j++){
                    temp[j] = scanner.nextInt();
                }
                list.add(temp);
            }
            boolean[] isVisited = new boolean[n];
            Arrays.fill(isVisited,false);
            dfs(list,isVisited,n,0,0);
            if(flag){
                System.out.println("YES");
            }else{
                System.out.println("NO");
            }
        }
    }
}